# ruff: noqa
# fmt: off
# type: ignore

## case: test_emit_errors

from algopy import arc4, subroutine

class Event(arc4.Struct):
    a: arc4.String
    b: arc4.UInt8

class Event2(arc4.Struct):
    a: arc4.String

@subroutine
def typed_ok() -> None:
    arc4.emit(Event(arc4.String("a"), arc4.UInt8(1)))
    arc4.emit("Event(string,uint8)", "a", 1)
    arc4.emit("Event(string,uint8)", "a", 255)

@subroutine
def abi_call_tuple_ok() -> None:
    event, txn = arc4.abi_call[Event]("foo()(string,uint8)")

@subroutine
def abi_call_wrong_return() -> None:
    event, txn = arc4.abi_call[Event]("foo()(string,uint64)") ## E: Method selector from args 'foo()(string,uint8)' does not match provided method selector: 'foo()(string,uint64)'

@subroutine
def wrong_arg_type1() -> None:
    arc4.emit("Event(string,uint64)", arc4.String("a"), arc4.UInt8(1)) ## E: Expected type algopy.arc4.UInt64, got type algopy.arc4.UInt8

@subroutine
def wrong_arg_type2() -> None:
    arc4.emit("Event(string,uint8)", b"a", 1) ## E: Can't construct algopy.arc4.String from Python literal b'a'

@subroutine
def wrong_arg_type3() -> None:
    arc4.emit("Event(string,uint8)", "a", 256) ## E: Invalid algopy.arc4.UInt8 value: 256

@subroutine
def wrong_arg_type4() -> None:
    arc4.emit("Event(string,uint8)", Event(arc4.String("a"), arc4.UInt8(1))) ## E: Number of arguments (1) does not match signature (2)

@subroutine
def wrong_arg_type5() -> None:
    arc4.emit("Event2(string)", Event2(arc4.String("a"))) ## E: Expected type algopy.arc4.String, got type test_emit_errors.Event2

@subroutine
def wrong_number_args() -> None:
    arc4.emit("Event(string,uint8)", "a", 1, 2) ## E: Number of arguments (3) does not match signature (2)

@subroutine
def invalid_signature1() -> None:
    arc4.emit("(string,uint8)", "a", 1) ## E: Invalid signature: name=''

@subroutine
def invalid_signature2() -> None:
    arc4.emit("string,uint8", "a", 1) ## E: Invalid signature: name='string,uint8'

@subroutine
def invalid_signature3() -> None:
    arc4.emit("9bad_name(string,uint8)", "a", 1) ## E: Invalid signature: name='9bad_name'

@subroutine
def invalid_signature4() -> None:
    arc4.emit("Event(string,uint8)void", arc4.String("a"), arc4.UInt8(1)) ## E: Invalid signature, trailing text after args 'void'

@subroutine
def invalid_signature5() -> None:
    arc4.emit("Event(string,uint8)uint64", arc4.String("a"), arc4.UInt8(1)) ## E: Invalid signature, trailing text after args 'uint64'

@subroutine
def invalid_signature6() -> None:
    arc4.emit("Event(string,uint8)(string,uint64)", arc4.String("a"), arc4.UInt8(1)) ## E: Invalid signature, trailing text after args '(string,uint64)'


## case: test_bad_uint_sizes

import typing

from algopy import arc4, subroutine

A: typing.TypeAlias = arc4.UIntN[typing.Literal[72]]
B: typing.TypeAlias = arc4.BigUIntN[typing.Literal[64]]

C: typing.TypeAlias = arc4.UFixedNxM[typing.Literal[72], typing.Literal[10]]
D: typing.TypeAlias = arc4.BigUFixedNxM[typing.Literal[64], typing.Literal[10]]


@subroutine
def testA(x: A) -> None: ## E: UIntN scale must be >=8 and <=64 bits, and be a multiple of 8
    assert x

@subroutine
def testB(x: B) -> None: ## E: BigUIntN scale must be >64 and <=512 bits, and be a multiple of 8
    assert x

@subroutine
def testC(x: C) -> None: ## E: UFixedNxM scale must be >=8 and <=64 bits, and be a multiple of 8
    assert x

@subroutine
def testD(x: D) -> None: ## E: BigUFixedNxM scale must be >64 and <=512 bits, and be a multiple of 8
    assert x


## case: test_arc4_tuple_immutable
from algopy import arc4

class NotOkay(arc4.ARC4Contract):
    @arc4.abimethod
    def not_okay(self) -> None:
        t = arc4.Tuple((arc4.UInt64(1), arc4.UInt8(1)))
        t[0] = arc4.UInt64(2)  # type: ignore ## E: expression is not valid as assignment target (algopy.arc4.Tuple[algopy.arc4.UInt64, algopy.arc4.UInt8] is immutable)


## case: arc4_copy

import typing
from algopy import arc4, GlobalState, subroutine

IntArray: typing.TypeAlias = arc4.DynamicArray[arc4.UInt64]


class Arc4CopyContract(arc4.ARC4Contract):
    def __init__(self) -> None:
        self.global_a = IntArray()
        self.global_b = GlobalState(IntArray())

    @subroutine
    def method_a(self, a: IntArray) -> None:
        a.append(arc4.UInt64(5))

    @arc4.abimethod
    def test(self) -> None:
        # Local does not need copy
        local_array = IntArray()
        self.method_a(local_array)

        # State vars should require copy

        self.method_a(self.global_a) ## E: algopy.arc4.DynamicArray[algopy.arc4.UInt64] must be copied using .copy() when being passed to a subroutine from state
        self.method_a(self.global_b.value) ## E: algopy.arc4.DynamicArray[algopy.arc4.UInt64] must be copied using .copy() when being passed to a subroutine from state

        # Require copy when building a tuple

        my_tuple = (local_array, local_array) ## E: algopy.arc4.DynamicArray[algopy.arc4.UInt64] must be copied using .copy() when being passed to a tuple expression

        # Don't allow destructuring of arc4 containers from tuples

        destructured_a, destructured_b = my_tuple ## E: Tuple cannot be destructured as it contains an item of type algopy.arc4.DynamicArray[algopy.arc4.UInt64] which requires copying. Use index access instead
        destructured_a = my_tuple[0] ## E: algopy.arc4.DynamicArray[algopy.arc4.UInt64] must be copied using .copy() when being assigned to another variable
        destructured_b = my_tuple[1] ## E: algopy.arc4.DynamicArray[algopy.arc4.UInt64] must be copied using .copy() when being assigned to another variable