# ruff: noqa
# fmt: off
# type: ignore

## case: test_typing_interactions

import typing

from algopy import Account, Bytes, UInt64, subroutine

SOME_ADDRESS = "VCMJKWOY5P5P7SKMZFFOCEROPJCZOTIJMNIYNUCKH7LRO45JMJP6UYBIJA"
SOME_ADDRESS_PADDED = SOME_ADDRESS + "=="


@subroutine
def typing_interactions() -> None:
    typing.reveal_type(UInt64(1)) ## N: Revealed type is "algopy._primitives.UInt64"
    typing.reveal_type((UInt64(1), Bytes(b""))) ## N: Revealed type is "tuple[algopy._primitives.UInt64, algopy._primitives.Bytes]"
    typing.assert_type(UInt64(1), UInt64) ## W: use of typing.assert_type has no effect on compilation \
                                          ## W: expression result is ignored

    assert typing.cast(Bytes, Account(SOME_ADDRESS)) == Bytes.from_base64(SOME_ADDRESS_PADDED) ## W: use of typing.cast, output may be invalid or insecure TEAL \
                                                                                               ## E: Unsupported comparison '==' between types

## case: test_unreachable

from algopy import Contract, UInt64, op


class ContractWithUnreachableCode(Contract):
    def approval_program(self) -> UInt64:
        x = UInt64(0)
        if x:
            return x
        else:
            op.err()
            return x ## E: Unreachable code

    def clear_state_program(self) -> bool:
        op.exit(1)
        return True ## E: Unreachable code

## case: test_invalid_address_literal

from algopy import arc4, Account, Bytes, Contract


class ContractWithInvalidAddressLiteral(Contract):
    def approval_program(self) -> bool:
        some_address = arc4.Address("ABC") ## E: Invalid address value. Address literals should be 58 characters and not include base32 padding
        return True

    def clear_state_program(self) -> bool:
        some_address = Account("ABC") ## E: Invalid address value. Address literals should be 58 characters and not include base32 padding
        return True

## case: test_undefined_variable

from algopy import Contract, Txn


class ContractWithUndefinedVariable(Contract):
    def approval_program(self) -> bool: ## E: Undefined register: undefined#1. This can be caused by attempting to reference variables that are only defined in other execution paths.
        if Txn.num_app_args:
            undefined = True
            return undefined
        return undefined

    def clear_state_program(self) -> bool:
        return True



## case: test_arc4_mutable_iteration

from algopy import *

@subroutine
def okay1() -> None:
    arr = arc4.DynamicBytes()
    for b in arr:
        assert b

@subroutine
def baddie1() -> None:
    arr_of_arr = arc4.DynamicArray[arc4.DynamicBytes]()
    for arr in arr_of_arr: ## E: Cannot directly iterate an ARC4 array of mutable objects, construct a for-loop over the indexes via urange(<array>.length) instead
        arr.append(arc4.Byte(1))


@subroutine
def okay2() -> None:
    arr_of_tup = arc4.DynamicArray[arc4.Tuple[arc4.UInt64, arc4.UInt64]]()
    for tup in arr_of_tup:
        assert tup[0] == tup[1]


@subroutine
def baddie2() -> None:
    arr_of_tup_with_arr = arc4.DynamicArray[arc4.Tuple[arc4.DynamicBytes, arc4.UInt64]]()
    for tup in arr_of_tup_with_arr: ## E: Cannot directly iterate an ARC4 array of mutable objects, construct a for-loop over the indexes via urange(<array>.length) instead
        tup[0].append(arc4.Byte(1))



## case: test_mutable_ref_param_reassignment_fails
from algopy import arc4

class Baddie(arc4.ARC4Contract):
    @arc4.abimethod
    def okay(self, arr: arc4.DynamicBytes) -> None:
        # this is allowed because nothing calls okay except for the router
        arr = arc4.DynamicBytes.from_bytes(arr.bytes)
        self.not_okay(arr)

    @arc4.abimethod()
    def not_okay(self, arr2: arc4.DynamicBytes) -> None:
        arr2 = arc4.DynamicBytes.from_bytes(arr2.bytes) ## E: Cannot reassign mutable parameter 'arr2' which is being passed by reference
        arr2.append(arc4.Byte(1))

## case: test_mutable_ref_param_reassignment_in_tuple_fails
from algopy import arc4

class Baddie(arc4.ARC4Contract):
    @arc4.abimethod
    def okay(self, arr: arc4.DynamicBytes) -> None:
        arr = arc4.DynamicBytes.from_bytes(arr.bytes)
        self.not_okay(arr)

    @arc4.abimethod()
    def not_okay(self, arr2: arc4.DynamicBytes) -> None:
        (arr2, foo) = (arc4.DynamicBytes.from_bytes(arr2.bytes), arc4.UInt64(1)) ## E: Cannot reassign mutable parameter 'arr2' which is being passed by reference
        arr2.append(arc4.Byte(1))
        assert foo


## case: test_state_proxies_outside_init_fail
from algopy import GlobalState, LocalState, Txn, UInt64, arc4

class Baddie(arc4.ARC4Contract):
    def __init__(self) -> None:
        self.global_a = GlobalState(UInt64)
        self.local_b = LocalState(UInt64)

    @arc4.abimethod
    def okay(self) -> None:
        self.global_a.value = UInt64(123)
        self.local_b[Txn.sender] = UInt64(456)

    @arc4.abimethod()
    def not_okay1(self) -> None:
        self.global_c = GlobalState(UInt64) ## E: algopy.GlobalState can only be assigned to a member variable in the __init__ method

    @arc4.abimethod()
    def not_okay2(self) -> None:
        self.local_d = LocalState(UInt64) ## E: algopy.LocalState can only be assigned to a member variable in the __init__ method


## case: bad_intrinsic_call
from algopy import *

@subroutine
def baddie() -> None:
    result = op.setbit_bytes(UInt64(0), 0, 0, 1) # type: ignore[call-arg,arg-type] ## E: Too many positional arguments \
                                                                                   ## E: Invalid argument type "algopy.UInt64" for argument "a" when calling algopy.op.setbit_bytes
    result = op.setbit_bytes(UInt64(0), 0, 0, unknown=1) # type: ignore[call-arg,arg-type] ## E: Unexpected argument \
                                                                                           ## E: Invalid argument type "algopy.UInt64" for argument "a" when calling algopy.op.setbit_bytes
    result = op.setbit_bytes(0, 0, 0, unknown=1) # type: ignore[call-arg,arg-type] ## E: Unexpected argument \
                                                                                   ## E: Unhandled literal type 'int' for argument
    result = op.setbit_bytes(b"", 0) # type: ignore[call-arg] ## E: Missing expected argument c

    result2 = op.GITxn.application_args(UInt64(0), UInt64(0)) # type: ignore[arg-type] ## E: Argument must be a literal int value

    foo = arc4.arc4_signature(String("foo(uint64)uint64")) # type: ignore[arg-type] ## E: Unexpected args for algopy.arc4.arc4_signature

    result3 = op.GITxn.fee() # type: ignore[call-arg] ## E: Missing expected argument t


## case: non_existent_members
from algopy import *

@subroutine
def uint64() -> None:
    UInt64().foo # type: ignore[attr-defined] ## E: unrecognised member of algopy.UInt64: foo

@subroutine
def biguint() -> None:
    BigUInt().foo # type: ignore[attr-defined] ## E: unrecognised member of algopy.BigUInt: foo

@subroutine
def bytes_() -> None:
    Bytes().foo # type: ignore[attr-defined] ## E: unrecognised member of algopy.Bytes: foo

@subroutine
def string() -> None:
    String().foo # type: ignore[attr-defined] ## E: unrecognised member of algopy.String: foo

@subroutine
def account() -> None:
    Account().foo # type: ignore[attr-defined] ## E: unrecognised member of algopy.Account: foo

@subroutine
def asset() -> None:
    Asset().foo # type: ignore[attr-defined] ## E: unrecognised member of algopy.Asset: foo

@subroutine
def application() -> None:
    Application().foo # type: ignore[attr-defined] ## E: unrecognised member of algopy.Application: foo


## case: invalid_constructor_args
from algopy import *

@subroutine
def invalid_constructor_args() -> None:
    a = UInt64("a") # type: ignore[arg-type] ## E: Invalid/unhandled arguments
    b = BigUInt("b") # type: ignore[arg-type] ## E: Invalid/unhandled arguments
    c = Bytes(3) # type: ignore[arg-type] ## E: Invalid/unhandled arguments
    d = String(4) # type: ignore[arg-type] ## E: Invalid/unhandled arguments
    e = Account(5) # type: ignore[arg-type] ## E: Invalid/unhandled arguments
    f = Asset("f") # type: ignore[arg-type] ## E: Invalid/unhandled arguments
    g = Application("g") # type: ignore[arg-type] ## E: Invalid/unhandled arguments
    # the next check is just to make sure we didn't throw on the line above

    h = 1 ## E: A Python literal is not valid at this location


## case: unsupported_type_comparisons
from algopy import *

@subroutine
def compare_biguint_string() -> None:
    assert BigUInt() != String() # type: ignore[operator,misc] ## E: Unsupported comparison '!=' between types


## case: float_div
from algopy import *

@subroutine
def float_div() -> None:
    assert UInt64(10) / 2 # type: ignore[operator] ## E: To maintain semantic compatibility with Python, only the truncating division operator (//) is supported
    assert BigUInt(5) / 2 # type: ignore[operator] ## E: To maintain semantic compatibility with Python, only the truncating division operator (//) is supported

    x = 1 ## E: A Python literal is not valid at this location


## case: unsupported_math_operators
from algopy import *

@subroutine
def unsupported_biguint_operator() -> None:
    assert BigUInt() ** 2 # type: ignore[operator,misc] ## E: Unsupported BigUInt math operator '**'

@subroutine
def unsupported_uint64_operator() -> None:
    assert UInt64() @ 2 # type: ignore[operator,misc] ## E: Unsupported UInt64 math operator '@'


## case: super_with_data_member
from algopy import *

class Base(ARC4Contract):
    def __init__(self) -> None:
        self.foo = UInt64(1)


class Derived(Base):
    def __init__(self) -> None:
        self.foo2 = 2 * super().foo ## E: unsupported reference to non-function member 'foo'


## case: redefinition
from algopy import *

class Base(ARC4Contract):
    def __init__(self) -> None:
        self.foo = UInt64(1)
        self.bar = UInt64(2)
        self.foo_p = GlobalState(UInt64(1))  ## N: Previous definition of foo_p was here
        self.bar_p = GlobalState(UInt64(2))  ## N: Previous definition of bar_p was here

class Derived1(Base):
    def __init__(self) -> None:
        super().__init__()
        self.foo = UInt64(2) # ok

class Derived2(Base):
    def __init__(self) -> None:
        super().__init__()
        self.bar = Bytes(b"bbb") # type: ignore[assignment] ## E: assignment target type differs from expression value type

class Derived3(Base):
    def __init__(self) -> None:
        super().__init__()
        self.foo_p = GlobalState(UInt64(3)) ## E: Redefinition of foo_p

class Derived4(Base):
    def __init__(self) -> None:
        super().__init__()
        self.bar_p = GlobalState(Bytes(b"bbb")) # type: ignore[arg-type] ## E: Redefinition of bar_p

class Base2(ARC4Contract):
    def __init__(self) -> None:
        if Txn.num_app_args:
            self.foo = UInt64(1)
            self.foo_p = GlobalState(UInt64(1)) ## N: Previous definition of foo_p was here
        else:
            self.foo = UInt64(2)
            self.foo_p = GlobalState(UInt64(2)) ## E: Redefinition of foo_p


## case: void_state
from algopy import *

class Con(Contract):
    x: None ## E: None is not supported as a value, only a return type

    def approval_program(self) -> bool:
        self.x = log(b"")  # type: ignore[func-returns-value] ## E: None indicates an empty return and cannot be assigned
        return True

    def clear_state_program(self) -> bool:
        assert self.x ## W: expression is always False
        return True
